packages/core/src/defer/triggering.ts

```diff
@@ -381,31 +381,70 @@ export async function triggerHydrationFromBlockName(
     return;
   }
 
-  // The parent promise is the possible case of a list of defer blocks already being queued
-  // If it is queued, it'll exist; otherwise it'll be null. The hydration queue will contain all
-  // elements that need to be hydrated, sans any that have promises already
+  // Trigger resource loading and hydration for the blocks in the queue in the order of highest block
+  // to lowest block. Once a block has finished resource loading, after next render fires after hydration
+  // finishes. The new block will have its defer instruction called and will be in the registry.
+  // Due to timing related to potential nested control flow, this has to be scheduled after the next render.
   const {parentBlockPromise, hydrationQueue} = getParentBlockHydrationQueue(blockName, injector);
+  if (hydrationQueue.length === 0) return;
+
+  // It's possible that the hydrationQueue topmost item is actually in the process of hydrating and has
+  // a promise already. In that case, we don't want to destroy that promise and queue it again.
+  if (parentBlockPromise !== null) {
+    hydrationQueue.shift();
+  }
 
   // The hydrating map in the registry prevents re-triggering hydration for a block that's already in
   // the hydration queue. Here we generate promises for each of the blocks about to be hydrated
   populateHydratingStateForQueue(dehydratedBlockRegistry, hydrationQueue);
 
-  // Trigger resource loading and hydration for the blocks in the queue in the order of highest block
-  // to lowest block. Once a block has finished resource loading, after next render fires after hydration
-  // finishes. The new block will have its defer instruction called and will be in the registry.
-  // Due to timing related to potential nested control flow, this has to be scheduled after the next render.
+  // We await this after populating the hydration state so we can prevent re-triggering hydration for
+  // the same blocks while this promise is being awaited.
+  if (parentBlockPromise !== null) {
+    await parentBlockPromise;
+  }
+
+  if (dehydratedBlockRegistry.has(hydrationQueue[0])) {
+    // the topmost parent block is already in the registry and we can proceed
+    // with hydration.
+    triggerHydrationForBlockQueue(injector, hydrationQueue, replayQueuedEventsFn);
+  } else {
+    // the topmost parent block is not yet in the registry, which may mean
+    // a lazy loaded route, a control flow branch was taken, a route has
+    // been navigated, etc. So we need to queue up the hydration process
+    // so that it can be finished after the top block has had its defer
+    // instruction executed.
+    const parentBlockAwaitCallbacks =
+      dehydratedBlockRegistry.awaitingCallbacks.get(hydrationQueue[0]) ?? [];
+    parentBlockAwaitCallbacks.push(() =>
+      triggerHydrationForBlockQueue(injector, hydrationQueue, replayQueuedEventsFn),
+    );
+    dehydratedBlockRegistry.awaitingCallbacks.set(hydrationQueue[0], parentBlockAwaitCallbacks);
```

Comment:
nit: for better code compression we can extract `awaitingCallbacks` to a const:
```suggestion
    const {awaitingCallbacks} = dehydratedBlockRegistry;
    const parentBlockAwaitCallbacks = awaitingCallbacks.get(hydrationQueue[0]) ?? [];
    parentBlockAwaitCallbacks.push(() =>
      triggerHydrationForBlockQueue(injector, hydrationQueue, replayQueuedEventsFn),
    );
    awaitingCallbacks.set(hydrationQueue[0], parentBlockAwaitCallbacks);
```

I was also wondering if this code can be in the registry class as well (thus making the `awaitingCallbacks` private), so the only code here would look something like this:

```
dehydratedBlockRegistry.awaitParentBlock(topmostParentBlock, () =>
  triggerHydrationForBlockQueue(injector, hydrationQueue, replayQueuedEventsFn)
);
```